sql/log_event_push_cond.cc

```diff
@@ -336,36 +336,94 @@ filed_str_to_sql_conditions(std::string &str_condition, const uchar *ptr,
   return 0;
 }
 
+//Check whether there are numbers or time types in the table
+static bool checklist_have_number_or_time(Field **field, uint field_num)
+{
+  for (uint col_id = 0; col_id < field_num; col_id++) {
+    Field *f = field[col_id];
+    switch (f->type()) {
+      case MYSQL_TYPE_LONG:
+      case MYSQL_TYPE_TINY:
+      case MYSQL_TYPE_INT24:
+      case MYSQL_TYPE_SHORT:
+      case MYSQL_TYPE_LONGLONG:
+      case MYSQL_TYPE_DOUBLE:
+      case MYSQL_TYPE_TIMESTAMP:
+      case MYSQL_TYPE_TIMESTAMP2:
+      case MYSQL_TYPE_DATETIME:
+      case MYSQL_TYPE_DATETIME2:
+      case MYSQL_TYPE_YEAR:
+      case MYSQL_TYPE_DATE:
+      case MYSQL_TYPE_NEWDATE:
+      case MYSQL_TYPE_TIME:
+      case MYSQL_TYPE_TIME2:
+      {
+        return true;
+      }
+      default:
+        break;
+    } 
+  }
+  return false;
+}
+
 
 bool Rows_log_event::column_information_to_conditions(std::string &sql_statemens, 
-                                                      MY_BITMAP *cols_bitmap, 
                                                       std::string &prefix)
 {
-  /*
-    Skip metadata bytes which gives the information about nullabity of master
-    columns. Master writes one bit for each affected column.
-   */
-  bitmap_bits_set(cols_bitmap);
   sql_statemens += prefix;
   Field **field = m_table->field;
-  MYSQL_TIME my_time;
-  std::memset(&my_time, 0, sizeof(my_time));
+  /*
+    If there is a unique constraint, 
+    use the field of the unique constraint as the push down condition
+  */
+  std::string key_field_name;
+  if(m_table->s->key_info && 
+    m_table->s->key_info->key_part &&
+    m_table->s->key_info->key_part->field)
+  {
+    key_field_name = m_table->s->key_info->key_part->field->field_name;
+  }
   int cond_num = 0;
+  bool unwanted_str = false;
+  /*
+    If there is a number or time type in the table, 
+    the string type value is not used as the push down condition.
+  */
+  if(checklist_have_number_or_time(field, m_table->s->fields)) unwanted_str = true;
+
   for (uint col_id = 0; col_id < m_table->s->fields; col_id++) {
     Field *f = field[col_id];
-    if (bitmap_is_set(cols_bitmap, col_id) == 0)
-      continue;
     uint16 meta;
     f->save_field_metadata((uchar*)&meta);
     std::string str_cond;
-    size_t size = filed_str_to_sql_conditions(str_cond, (f->is_null() ? NULL: f->ptr), f->type(), meta, f);
+
+    if(!key_field_name.empty()){
```

Comment:
if `key_field_name` is null, whether we continue or not? I think this part can be refined.